iree/turbine/kernel/wave/scheduling/multi_buffering.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+from ..._support.tracing import CapturedTrace
+from ..._support.indexing import (
+    IndexSymbol,
+    xor,
+)
+from ...compiler.base import CodegenError
+from ...lang.global_symbols import SHARED_ADDRESS_SPACE
+from ...ops.wave_ops import (
+    get_custom,
+    Read,
+    Write,
+    CustomOp,
+    Reduction,
+)
+import iree.turbine.kernel.lang as tkl
+
+
+def multi_buffer(trace: CapturedTrace):
+    """Perform multi buffering for all supported shared memory locations"""
+
+    # Find all reductions
+    reductions = trace.walk(lambda node: isinstance(get_custom(node), Reduction))
+
+    # Get reduction dimension from first reduction
+    if not reductions or len(reductions) != 1:
+        raise CodegenError(
+            f"Unexpected number of reductions found in graph: {len(reductions)} vs 1"
+        )
+
+    reduction_axis = get_custom(reductions[0]).axis
+
+    # Find reads that index using the reduction dim and are from shared memory
```

Comment:
Why are you looking for reads that index using the reduction dim? Is it so that you only capture the operands to the GEMM and not the outputs? If so, Is there anyway you can relax this restriction as it would not work for attention (if we promote the query matrix)?

Reply:
I lifted this restriction as well. 👍 
There is currently still a bug when targeting attention. I think we can still proceed with the PR and fix this in a follow-up.